File: backend/app/modules/litigation_approvals_compliance/compute.py
"""Derived model for Litigation, Approvals & Compliance."""

from __future__ import annotations

from datetime import UTC, datetime
from math import ceil
from typing import Any

from app.modules.litigation_approvals_compliance.approvals import get_approvals, is_perpetual_approval
from app.modules.litigation_approvals_compliance.constants import RECONCILIATION_TOLERANCE
from app.modules.litigation_approvals_compliance.decimal_utils import (
    add_decimals,
    days_between_dates,
    is_filled_decimal,
    parse_decimal,
    parse_iso_date,
    subtract_decimals,
)
from app.modules.litigation_approvals_compliance.matters import (
    get_matters,
    is_criminal_matter,
    is_tax_matter,
)


def _currency_key(matter: dict[str, Any]) -> str:
    amounts = matter.get("amounts") or {}
    currency = str(amounts.get("currency") or "").strip() or "UNSPECIFIED"
    unit = str(amounts.get("amountUnit") or "").strip() or "unspecified"
    return f"{currency}::{unit}"


def _matter_exposure(matter: dict[str, Any]) -> str:
    amounts = matter.get("amounts") or {}
    total = str(amounts.get("totalQuantifiedAmount") or "")
    if is_filled_decimal(total):
        return total
    return add_decimals(
        str(amounts.get("principalClaim") or ""),
        str(amounts.get("taxDemand") or ""),
        str(amounts.get("interest") or ""),
        str(amounts.get("penalty") or ""),
        str(amounts.get("fine") or ""),
        str(amounts.get("damages") or ""),
        str(amounts.get("compensation") or ""),
        str(amounts.get("otherExposure") or ""),
    )


def _build_exposure_by_currency(matters: list[dict[str, Any]]) -> list[dict[str, Any]]:
    groups: dict[str, list[dict[str, Any]]] = {}
    for matter in matters:
        key = _currency_key(matter)
        groups.setdefault(key, []).append(matter)

    result: list[dict[str, Any]] = []
    for key, group in groups.items():
        currency, amount_unit = key.split("::", 1)
        tax_matters = [m for m in group if is_tax_matter(m)]
        result.append(
            {
                "currency": currency,
                "amountUnit": amount_unit,
                "matterCount": len(group),
                "totalExposure": add_decimals(*[_matter_exposure(m) for m in group]),
                "taxExposure": add_decimals(
                    *[
                        str((m.get("amounts") or {}).get("taxDemand") or "") or _matter_exposure(m)
                        for m in tax_matters
                    ]
                ),
                "criminalCount": sum(1 for m in group if is_criminal_matter(m)),
                "pendingCount": sum(
                    1
                    for m in group
                    if (m.get("statusOutcome") or {}).get("outcomeStatus")
                    in {"pending", "appeal-pending", ""}
                ),
            }
        )
    return result


def _build_matters_by_category(matters: list[dict[str, Any]]) -> list[dict[str, Any]]:
    counts: dict[str, int] = {}
    for matter in matters:
        category = str((matter.get("identity") or {}).get("category") or "") or "unspecified"
        counts[category] = counts.get(category, 0) + 1
    return sorted(
        [{"category": category, "count": count} for category, count in counts.items()],
        key=lambda item: item["count"],
        reverse=True,
    )


def _build_tax_aggregates(payload: dict[str, Any]) -> dict[str, Any]:
    section3 = payload.get("criminalRegulatoryTaxAndEnforcementReadiness") or {}
    tax_details = [
        detail for detail in (section3.get("taxProceedingDetails") or []) if isinstance(detail, dict)
    ]
    direct_types = {"direct-tax"}
    indirect_types = {"gst", "customs", "excise", "vat-sales-tax", "service-tax"}

    direct_tax_demand = ""
    indirect_tax_demand = ""

    for detail in tax_details:
        demand = add_decimals(
            str(detail.get("demand") or ""),
            str(detail.get("interest") or ""),
            str(detail.get("penalty") or ""),
        )
        tax_type = str(detail.get("taxType") or "")
        if tax_type in direct_types:
            direct_tax_demand = add_decimals(direct_tax_demand, demand)
        elif tax_type in indirect_types:
            indirect_tax_demand = add_decimals(indirect_tax_demand, demand)
        elif tax_type:
            direct_tax_demand = add_decimals(direct_tax_demand, demand)

    for matter in get_matters(payload):
        if not is_tax_matter(matter):
            continue
        amounts = matter.get("amounts") or {}
        demand = str(amounts.get("taxDemand") or "") or _matter_exposure(matter)
        if (matter.get("identity") or {}).get("category") == "tax":
            direct_tax_demand = add_decimals(direct_tax_demand, demand)

    return {
        "directTaxDemand": direct_tax_demand,
        "indirectTaxDemand": indirect_tax_demand,
        "totalDemand": add_decimals(direct_tax_demand, indirect_tax_demand),
        "totalBalanceDisputed": add_decimals(
            *[str(detail.get("balanceDisputed") or "") for detail in tax_details]
        ),
        "proceedingCount": len(tax_details),
    }


def _build_approval_expiry_windows(
    payload: dict[str, Any],
    as_of: datetime,
) -> dict[str, list[dict[str, Any]]]:
    windows: dict[str, list[dict[str, Any]]] = {
        "within30Days": [],
        "within90Days": [],
        "within180Days": [],
        "within365Days": [],
    }

    from datetime import timedelta

    horizon365 = as_of + timedelta(days=365)

    for approval in get_approvals(payload):
        if is_perpetual_approval(approval):
            continue
        details = approval.get("details") or {}
        renewal = approval.get("renewalMetadata") or {}
        expiry = parse_iso_date(str(details.get("expiryDate") or "")) or parse_iso_date(
            str(renewal.get("renewalDueDate") or "")
        )
        if expiry is None or expiry > horizon365:
            continue

        days_until = ceil((expiry - as_of).total_seconds() / (60 * 60 * 24))
        identity = approval.get("identity") or {}
        holder = approval.get("holder") or {}
        label = (
            str(identity.get("approvalLicenceName") or "").strip()
            or str(holder.get("displayName") or "").strip()
            or str(approval.get("approvalId") or "")[:8]
        )
        expiry_date = str(details.get("expiryDate") or "").strip() or str(
            renewal.get("renewalDueDate") or ""
        ).strip()
        entry_base = {
            "approvalId": str(approval.get("approvalId") or ""),
            "label": label,
            "expiryDate": expiry_date,
            "daysUntilExpiry": days_until,
        }

        if days_until <= 30:
            windows["within30Days"].append({**entry_base, "window": "30"})
        if days_until <= 90:
            windows["within90Days"].append({**entry_base, "window": "90"})
        if days_until <= 180:
            windows["within180Days"].append({**entry_base, "window": "180"})
        if days_until <= 365:
            windows["within365Days"].append({**entry_base, "window": "365"})

    return windows


def _reconciliation_status_label(status: str, linked_available: bool) -> str:
    if not linked_available:
        return "Pending linked workstream"
    mapping = {
        "reconciled": "Reconciled",
        "potential-inconsistency": "Potential inconsistency",
        "pending-professional-confirmation": "Pending professional confirmation",
        "pending-linked-workstream": "Pending linked workstream",
        "missing-information": "Missing information",
    }
    if status in mapping:
        return mapping[status]
    return status.replace("-", " ") if status else "Not captured"


def _build_reconciliation_preview(
    payload: dict[str, Any],
    linked_references: dict[str, Any],
) -> dict[str, Any]:
    reconciliation = payload.get("reconciliationRemediationAndIssuerConfirmations") or {}
    financials_rec = reconciliation.get("financialsReconciliation") or {}
    tax_aggregates = _build_tax_aggregates(payload)
    matters = get_matters(payload)
    litigation_aggregate = str(financials_rec.get("litigationAggregateAmount") or "") or add_decimals(
        *[_matter_exposure(m) for m in matters]
    )

    financials_kpis = linked_references.get("financialsKpis") or {}
    litigation_difference = str(financials_rec.get("litigationDifference") or "")
    contingent = financials_kpis.get("contingentLiabilitiesTotal")
    if not is_filled_decimal(litigation_difference) and financials_kpis.get("available") and contingent:
        litigation_difference = subtract_decimals(litigation_aggregate, str(contingent))

    financials_status = str(financials_rec.get("reconciliationStatus") or "")
    if not financials_status and financials_kpis.get("available") and contingent:
        diff = abs(parse_decimal(litigation_difference) or 0)
        financials_status = (
            "reconciled" if diff <= RECONCILIATION_TOLERANCE else "potential-inconsistency"
        )

    return {
        "financials": {
            "status": _reconciliation_status_label(
                financials_status,
                bool(financials_kpis.get("available")),
            ),
            "detail": (
                f"Litigation aggregate {litigation_aggregate or '—'} vs Financials contingent liabilities {contingent or '—'}. Tax aggregate {tax_aggregates.get('totalDemand') or '—'}."
                if financials_kpis.get("available")
                else "Financials & KPIs linked data not yet available."
            ),
        },
        "groupEntities": {
            "status": _reconciliation_status_label(
                str((reconciliation.get("groupEntitiesReconciliation") or {}).get("reconciliationStatus") or ""),
                bool((linked_references.get("groupEntities") or {}).get("available")),
            ),
            "detail": (
                "Group Entities legal declarations and subsidiary litigation cross-check."
                if (linked_references.get("groupEntities") or {}).get("available")
                else "Group Entities linked data not yet available."
            ),
        },
        "managementGovernance": {
            "status": _reconciliation_status_label(
                str(
                    (reconciliation.get("managementGovernanceReconciliation") or {}).get(
                        "reconciliationStatus"
                    )
                    or ""
                ),
                bool((linked_references.get("managementGovernance") or {}).get("available")),
            ),
            "detail": (
                "Management & Governance director/KMP legal declarations cross-check."
                if (linked_references.get("managementGovernance") or {}).get("available")
                else "Management & Governance linked data not yet available."
            ),
        },
        "bac": {
            "status": _reconciliation_status_label(
                str((reconciliation.get("bacReconciliation") or {}).get("reconciliationStatus") or ""),
                bool((linked_references.get("borrowingsAssetsContracts") or {}).get("available")),
            ),
            "detail": (
                "Borrowings, Assets & Contracts defaults, disputes and lender matters cross-check."
                if (linked_references.get("borrowingsAssetsContracts") or {}).get("available")
                else "Borrowings, Assets & Contracts linked data not yet available."
            ),
        },
        "businessOperations": {
            "status": _reconciliation_status_label(
                str(
                    (reconciliation.get("businessOperationsReconciliation") or {}).get(
                        "reconciliationStatus"
                    )
                    or ""
                ),
                bool((linked_references.get("businessOperations") or {}).get("available")),
            ),
            "detail": (
                "Business & Operations facilities, licences and operational incidents cross-check."
                if (linked_references.get("businessOperations") or {}).get("available")
                else "Business & Operations linked data not yet available."
            ),
        },
        "objectsOfIssue": {
            "status": _reconciliation_status_label(
                str(
                    (reconciliation.get("objectsOfIssueReconciliation") or {}).get(
                        "reconciliationStatus"
                    )
                    or ""
                ),
                bool((linked_references.get("objectsOfIssue") or {}).get("available")),
            ),
            "detail": (
                "Objects of the Issue capex/expansion approval plan cross-check."
                if (linked_references.get("objectsOfIssue") or {}).get("available")
                else "Objects of the Issue linked data not yet available."
            ),
        },
        "ipoSetup": {
            "status": _reconciliation_status_label(
                str((reconciliation.get("ipoSetupReconciliation") or {}).get("reconciliationStatus") or ""),
                bool((linked_references.get("ipoSetup") or {}).get("available")),
            ),
            "detail": (
                "IPO Setup & Eligibility debarment and serious proceedings declarations cross-check."
                if (linked_references.get("ipoSetup") or {}).get("available")
                else "IPO Setup & Eligibility linked data not yet available."
            ),
        },
    }


def _build_creditor_totals(payload: dict[str, Any]) -> dict[str, Any]:
    section = payload.get("materialCreditorsPenaltiesAndMaterialDevelopments") or {}
    aggregates = section.get("creditorAggregateInputs") or {}
    creditors = [c for c in (section.get("materialCreditors") or []) if isinstance(c, dict)]

    material_outstanding = str(aggregates.get("materialCreditorAmount") or "") or add_decimals(
        *[
            str(c.get("amountOutstanding") or "")
            for c in creditors
            if c.get("msmeStatus") != "yes"
        ]
    )
    msme_outstanding = str(aggregates.get("msmeOutstandingAmount") or "") or add_decimals(
        *[str(c.get("amountOutstanding") or "") for c in creditors if c.get("msmeStatus") == "yes"]
    )

    return {
        "materialCreditorCount": sum(1 for c in creditors if c.get("msmeStatus") != "yes"),
        "msmeCreditorCount": sum(1 for c in creditors if c.get("msmeStatus") == "yes"),
        "materialOutstanding": material_outstanding,
        "msmeOutstanding": msme_outstanding,
        "aggregateOutstanding": add_decimals(material_outstanding, msme_outstanding),
        "reconciliationDifference": str(aggregates.get("reconciliationDifference") or ""),
        "reconciliationStatus": str(aggregates.get("reconciliationStatus") or ""),
    }


def compute_litigation_approvals_compliance_model(
    payload: dict[str, Any],
    linked_references: dict[str, Any],
) -> dict[str, Any]:
    matters = get_matters(payload)
    approvals = get_approvals(payload)
    section6 = payload.get("corporateStatutoryAndOperationalComplianceExceptions") or {}
    section5 = payload.get("approvalConditionsFacilityComplianceAndRenewalReadiness") or {}
    as_of = datetime.now(tz=UTC)

    delayed_statutory_dues = 0
    for due in section6.get("statutoryDues") or []:
        if not isinstance(due, dict):
            continue
        delay = parse_decimal(str(due.get("delayDays") or ""))
        if delay is None and due.get("dueDate") and due.get("paymentDate"):
            delay = days_between_dates(str(due.get("dueDate")), str(due.get("paymentDate")))
        if delay is not None and delay > 0:
            delayed_statutory_dues += 1

    approval_conditions_outstanding = sum(
        1
        for condition in (section5.get("approvalConditions") or [])
        if isinstance(condition, dict)
        and condition.get("complianceStatus") in {"pending", "delayed", "not-sure"}
    )

    snapshot = (payload.get("legalUniverseMaterialityPolicyAndPartyMapping") or {}).get(
        "legalDdSnapshot"
    ) or {}

    return {
        "matterCount": len(matters),
        "mattersByCategory": _build_matters_by_category(matters),
        "criminalMatterCount": sum(1 for m in matters if is_criminal_matter(m)),
        "taxMatterCount": sum(1 for m in matters if is_tax_matter(m)),
        "pendingOutcomeCount": sum(
            1
            for m in matters
            if (m.get("statusOutcome") or {}).get("outcomeStatus")
            in {"pending", "appeal-pending", ""}
        ),
        "exposureByCurrency": _build_exposure_by_currency(matters),
        "taxAggregates": _build_tax_aggregates(payload),
        "approvalCount": len(approvals),
        "expiredApprovalCount": sum(
            1
            for a in approvals
            if a.get("status") in {"expired-renewal-applied", "expired-renewal-not-applied"}
        ),
        "renewalPendingCount": sum(
            1 for a in approvals if a.get("status") in {"renewal-pending", "application-pending"}
        ),
        "approvalExpiryWindows": _build_approval_expiry_windows(payload, as_of),
        "complianceCounts": {
            "domainReviewCount": len(section6.get("complianceDomainReviews") or []),
            "domainsWithKnownExceptions": sum(
                1
                for review in (section6.get("complianceDomainReviews") or [])
                if isinstance(review, dict) and review.get("knownExceptions") == "yes"
            ),
            "complianceIssueCount": len(section6.get("complianceIssues") or []),
            "continuingIssues": sum(
                1
                for issue in (section6.get("complianceIssues") or [])
                if isinstance(issue, dict) and issue.get("continuing") == "yes"
            ),
            "statutoryDueCount": len(section6.get("statutoryDues") or []),
            "delayedStatutoryDues": delayed_statutory_dues,
            "approvalConditionsOutstanding": approval_conditions_outstanding,
        },
        "creditorTotals": _build_creditor_totals(payload),
        "remediationOpenCount": sum(
            1
            for action in (
                (payload.get("reconciliationRemediationAndIssuerConfirmations") or {}).get(
                    "remediationActions"
                )
                or []
            )
            if isinstance(action, dict)
            and action.get("status") in {"open", "in-progress", "blocked"}
        ),
        "legalDdAsOfDate": str(snapshot.get("legalDdAsOfDate") or ""),
        "reconciliation": _build_reconciliation_preview(payload, linked_references),
    }

File: frontend/lib/litigation-approvals-compliance/compute.ts
/**
 * Derived model for Litigation, Approvals & Compliance (LAC1, frontend-only).
 */

import { getApprovals, isPerpetualApproval } from '@/lib/litigation-approvals-compliance/approvals';
import {
  addDecimals,
  daysBetweenDates,
  isFilledDecimal,
  parseDecimal,
  parseIsoDate,
  subtractDecimals,
} from '@/lib/litigation-approvals-compliance/decimal';
import { getMatters, isCriminalMatter, isTaxMatter } from '@/lib/litigation-approvals-compliance/matters';
import type { LinkedWorkstreamReferences } from '@/lib/litigation-approvals-compliance/types';
import type {
  LitigationApprovalsCompliancePayload,
  MatterCategory,
  MatterRecord,
} from '@/lib/schemas/litigation-approvals-compliance';

const RECONCILIATION_TOLERANCE = 1;

export type MatterCategoryCount = {
  category: MatterCategory | 'unspecified';
  count: number;
};

export type ExposureByCurrency = {
  currency: string;
  amountUnit: string;
  matterCount: number;
  totalExposure: string;
  taxExposure: string;
  criminalCount: number;
  pendingCount: number;
};

export type TaxAggregate = {
  directTaxDemand: string;
  indirectTaxDemand: string;
  totalDemand: string;
  totalBalanceDisputed: string;
  proceedingCount: number;
};

export type ApprovalExpiryWindowEntry = {
  approvalId: string;
  label: string;
  expiryDate: string;
  daysUntilExpiry: number | null;
  window: '30' | '90' | '180' | '365';
};

export type ApprovalExpiryWindows = {
  within30Days: ApprovalExpiryWindowEntry[];
  within90Days: ApprovalExpiryWindowEntry[];
  within180Days: ApprovalExpiryWindowEntry[];
  within365Days: ApprovalExpiryWindowEntry[];
};

export type ComplianceCounts = {
  domainReviewCount: number;
  domainsWithKnownExceptions: number;
  complianceIssueCount: number;
  continuingIssues: number;
  statutoryDueCount: number;
  delayedStatutoryDues: number;
  approvalConditionsOutstanding: number;
};

export type CreditorTotals = {
  materialCreditorCount: number;
  msmeCreditorCount: number;
  materialOutstanding: string;
  msmeOutstanding: string;
  aggregateOutstanding: string;
  reconciliationDifference: string;
  reconciliationStatus: string;
};

export type LacReconciliationPreview = {
  financials: { status: string; detail: string };
  groupEntities: { status: string; detail: string };
  managementGovernance: { status: string; detail: string };
  bac: { status: string; detail: string };
  businessOperations: { status: string; detail: string };
  objectsOfIssue: { status: string; detail: string };
  ipoSetup: { status: string; detail: string };
};

export type LitigationApprovalsComplianceModel = {
  matterCount: number;
  mattersByCategory: MatterCategoryCount[];
  criminalMatterCount: number;
  taxMatterCount: number;
  pendingOutcomeCount: number;
  exposureByCurrency: ExposureByCurrency[];
  taxAggregates: TaxAggregate;
  approvalCount: number;
  expiredApprovalCount: number;
  renewalPendingCount: number;
  approvalExpiryWindows: ApprovalExpiryWindows;
  complianceCounts: ComplianceCounts;
  creditorTotals: CreditorTotals;
  remediationOpenCount: number;
  legalDdAsOfDate: string;
  reconciliation: LacReconciliationPreview;
};

function currencyKey(matter: MatterRecord): string {
  const currency = matter.amounts.currency.trim() || 'UNSPECIFIED';
  const unit = matter.amounts.amountUnit.trim() || 'unspecified';
  return `${currency}::${unit}`;
}

function matterExposure(matter: MatterRecord): string {
  if (isFilledDecimal(matter.amounts.totalQuantifiedAmount)) {
    return matter.amounts.totalQuantifiedAmount;
  }
  return addDecimals(
    matter.amounts.principalClaim,
    matter.amounts.taxDemand,
    matter.amounts.interest,
    matter.amounts.penalty,
    matter.amounts.fine,
    matter.amounts.damages,
    matter.amounts.compensation,
    matter.amounts.otherExposure,
  );
}

function buildExposureByCurrency(matters: MatterRecord[]): ExposureByCurrency[] {
  const groups = new Map<string, MatterRecord[]>();
  for (const matter of matters) {
    const key = currencyKey(matter);
    const list = groups.get(key) ?? [];
    list.push(matter);
    groups.set(key, list);
  }

  return [...groups.entries()].map(([key, group]) => {
    const [currency, amountUnit] = key.split('::');
    const taxMatters = group.filter(isTaxMatter);
    return {
      currency,
      amountUnit,
      matterCount: group.length,
      totalExposure: addDecimals(...group.map(matterExposure)),
      taxExposure: addDecimals(...taxMatters.map((m) => m.amounts.taxDemand || matterExposure(m))),
      criminalCount: group.filter(isCriminalMatter).length,
      pendingCount: group.filter(
        (m) =>
          m.statusOutcome.outcomeStatus === 'pending' ||
          m.statusOutcome.outcomeStatus === 'appeal-pending' ||
          m.statusOutcome.outcomeStatus === '',
      ).length,
    };
  });
}

function buildMattersByCategory(matters: MatterRecord[]): MatterCategoryCount[] {
  const counts = new Map<MatterCategory | 'unspecified', number>();
  for (const matter of matters) {
    const category = matter.identity.category || 'unspecified';
    counts.set(category, (counts.get(category) ?? 0) + 1);
  }
  return [...counts.entries()]
    .map(([category, count]) => ({ category, count }))
    .sort((a, b) => b.count - a.count);
}

function buildTaxAggregates(payload: LitigationApprovalsCompliancePayload): TaxAggregate {
  const taxDetails = payload.criminalRegulatoryTaxAndEnforcementReadiness.taxProceedingDetails;
  const directTypes = new Set(['direct-tax']);
  const indirectTypes = new Set(['gst', 'customs', 'excise', 'vat-sales-tax', 'service-tax']);

  let directTaxDemand = '';
  let indirectTaxDemand = '';

  for (const detail of taxDetails) {
    const demand = addDecimals(detail.demand, detail.interest, detail.penalty);
    if (directTypes.has(detail.taxType)) {
      directTaxDemand = addDecimals(directTaxDemand, demand);
    } else if (indirectTypes.has(detail.taxType)) {
      indirectTaxDemand = addDecimals(indirectTaxDemand, demand);
    } else if (detail.taxType) {
      directTaxDemand = addDecimals(directTaxDemand, demand);
    }
  }

  for (const matter of getMatters(payload)) {
    if (!isTaxMatter(matter)) continue;
    const demand = matter.amounts.taxDemand || matterExposure(matter);
    if (matter.identity.category === 'tax') {
      directTaxDemand = addDecimals(directTaxDemand, demand);
    }
  }

  return {
    directTaxDemand,
    indirectTaxDemand,
    totalDemand: addDecimals(directTaxDemand, indirectTaxDemand),
    totalBalanceDisputed: addDecimals(
      ...taxDetails.map((detail) => detail.balanceDisputed),
    ),
    proceedingCount: taxDetails.length,
  };
}

function buildApprovalExpiryWindows(
  payload: LitigationApprovalsCompliancePayload,
  asOf: Date,
): ApprovalExpiryWindows {
  const windows: ApprovalExpiryWindows = {
    within30Days: [],
    within90Days: [],
    within180Days: [],
    within365Days: [],
  };

  const horizon365 = new Date(asOf);
  horizon365.setDate(horizon365.getDate() + 365);

  for (const approval of getApprovals(payload)) {
    if (isPerpetualApproval(approval)) continue;
    const expiry =
      parseIsoDate(approval.details.expiryDate) ??
      parseIsoDate(approval.renewalMetadata.renewalDueDate);
    if (!expiry || expiry > horizon365) continue;

    const daysUntil = Math.ceil((expiry.getTime() - asOf.getTime()) / (1000 * 60 * 60 * 24));
    const label =
      approval.identity.approvalLicenceName.trim() ||
      approval.holder.displayName.trim() ||
      approval.approvalId.slice(0, 8);
    const expiryDate =
      approval.details.expiryDate.trim() || approval.renewalMetadata.renewalDueDate.trim();

    const entryBase = {
      approvalId: approval.approvalId,
      label,
      expiryDate,
      daysUntilExpiry: daysUntil,
    };

    if (daysUntil <= 30) {
      windows.within30Days.push({ ...entryBase, window: '30' });
    }
    if (daysUntil <= 90) {
      windows.within90Days.push({ ...entryBase, window: '90' });
    }
    if (daysUntil <= 180) {
      windows.within180Days.push({ ...entryBase, window: '180' });
    }
    if (daysUntil <= 365) {
      windows.within365Days.push({ ...entryBase, window: '365' });
    }
  }

  return windows;
}

function reconciliationStatusLabel(status: string, linkedAvailable: boolean): string {
  if (!linkedAvailable) return 'Pending linked workstream';
  switch (status) {
    case 'reconciled':
      return 'Reconciled';
    case 'potential-inconsistency':
      return 'Potential inconsistency';
    case 'pending-professional-confirmation':
      return 'Pending professional confirmation';
    case 'pending-linked-workstream':
      return 'Pending linked workstream';
    case 'missing-information':
      return 'Missing information';
    default:
      return status ? status.replaceAll('-', ' ') : 'Not captured';
  }
}

function buildReconciliationPreview(
  payload: LitigationApprovalsCompliancePayload,
  linkedReferences: LinkedWorkstreamReferences,
): LacReconciliationPreview {
  const reconciliation = payload.reconciliationRemediationAndIssuerConfirmations;
  const financialsRec = reconciliation.financialsReconciliation;
  const taxAggregates = buildTaxAggregates(payload);
  const matters = getMatters(payload);
  const litigationAggregate =
    financialsRec.litigationAggregateAmount ||
    addDecimals(...matters.map(matterExposure));

  let litigationDifference = financialsRec.litigationDifference;
  if (
    !isFilledDecimal(litigationDifference) &&
    linkedReferences.financialsKpis.available &&
    linkedReferences.financialsKpis.contingentLiabilitiesTotal
  ) {
    litigationDifference = subtractDecimals(
      litigationAggregate,
      linkedReferences.financialsKpis.contingentLiabilitiesTotal,
    );
  }

  let financialsStatus = financialsRec.reconciliationStatus;
  if (!financialsStatus && linkedReferences.financialsKpis.available) {
    const diff = Math.abs(parseDecimal(litigationDifference) ?? 0);
    financialsStatus =
      diff <= RECONCILIATION_TOLERANCE ? 'reconciled' : 'potential-inconsistency';
  }

  return {
    financials: {
      status: reconciliationStatusLabel(
        financialsStatus,
        linkedReferences.financialsKpis.available,
      ),
      detail: linkedReferences.financialsKpis.available
        ? `Litigation aggregate ${litigationAggregate || '—'} vs Financials contingent liabilities ${linkedReferences.financialsKpis.contingentLiabilitiesTotal ?? '—'}. Tax aggregate ${taxAggregates.totalDemand || '—'}.`
        : 'Financials & KPIs linked data not yet available.',
    },
    groupEntities: {
      status: reconciliationStatusLabel(
        reconciliation.groupEntitiesReconciliation.reconciliationStatus,
        linkedReferences.groupEntities.available,
      ),
      detail: linkedReferences.groupEntities.available
        ? 'Group Entities legal declarations and subsidiary litigation cross-check.'
        : 'Group Entities linked data not yet available.',
    },
    managementGovernance: {
      status: reconciliationStatusLabel(
        reconciliation.managementGovernanceReconciliation.reconciliationStatus,
        linkedReferences.managementGovernance.available,
      ),
      detail: linkedReferences.managementGovernance.available
        ? 'Management & Governance director/KMP legal declarations cross-check.'
        : 'Management & Governance linked data not yet available.',
    },
    bac: {
      status: reconciliationStatusLabel(
        reconciliation.bacReconciliation.reconciliationStatus,
        linkedReferences.borrowingsAssetsContracts.available,
      ),
      detail: linkedReferences.borrowingsAssetsContracts.available
        ? 'Borrowings, Assets & Contracts defaults, disputes and lender matters cross-check.'
        : 'Borrowings, Assets & Contracts linked data not yet available.',
    },
    businessOperations: {
      status: reconciliationStatusLabel(
        reconciliation.businessOperationsReconciliation.reconciliationStatus,
        linkedReferences.businessOperations.available,
      ),
      detail: linkedReferences.businessOperations.available
        ? 'Business & Operations facilities, licences and operational incidents cross-check.'
        : 'Business & Operations linked data not yet available.',
    },
    objectsOfIssue: {
      status: reconciliationStatusLabel(
        reconciliation.objectsOfIssueReconciliation.reconciliationStatus,
        linkedReferences.objectsOfIssue.available,
      ),
      detail: linkedReferences.objectsOfIssue.available
        ? 'Objects of the Issue capex/expansion approval plan cross-check.'
        : 'Objects of the Issue linked data not yet available.',
    },
    ipoSetup: {
      status: reconciliationStatusLabel(
        reconciliation.ipoSetupReconciliation.reconciliationStatus,
        linkedReferences.ipoSetup.available,
      ),
      detail: linkedReferences.ipoSetup.available
        ? 'IPO Setup & Eligibility debarment and serious proceedings declarations cross-check.'
        : 'IPO Setup & Eligibility linked data not yet available.',
    },
  };
}

function buildCreditorTotals(payload: LitigationApprovalsCompliancePayload): CreditorTotals {
  const section = payload.materialCreditorsPenaltiesAndMaterialDevelopments;
  const aggregates = section.creditorAggregateInputs;
  const creditors = section.materialCreditors;

  const materialOutstanding =
    aggregates.materialCreditorAmount ||
    addDecimals(
      ...creditors
        .filter((c) => c.msmeStatus !== 'yes')
        .map((c) => c.amountOutstanding),
    );
  const msmeOutstanding =
    aggregates.msmeOutstandingAmount ||
    addDecimals(
      ...creditors.filter((c) => c.msmeStatus === 'yes').map((c) => c.amountOutstanding),
    );

  return {
    materialCreditorCount: creditors.filter((c) => c.msmeStatus !== 'yes').length,
    msmeCreditorCount: creditors.filter((c) => c.msmeStatus === 'yes').length,
    materialOutstanding,
    msmeOutstanding,
    aggregateOutstanding: addDecimals(materialOutstanding, msmeOutstanding),
    reconciliationDifference: aggregates.reconciliationDifference,
    reconciliationStatus: aggregates.reconciliationStatus,
  };
}

export function computeLitigationApprovalsComplianceModel(
  payload: LitigationApprovalsCompliancePayload,
  linkedReferences: LinkedWorkstreamReferences,
): LitigationApprovalsComplianceModel {
  const matters = getMatters(payload);
  const approvals = getApprovals(payload);
  const section6 = payload.corporateStatutoryAndOperationalComplianceExceptions;
  const section5 = payload.approvalConditionsFacilityComplianceAndRenewalReadiness;
  const asOf = new Date();

  const delayedStatutoryDues = section6.statutoryDues.filter((due) => {
    const delay =
      parseDecimal(due.delayDays) ??
      (due.dueDate && due.paymentDate ? daysBetweenDates(due.dueDate, due.paymentDate) : null);
    return delay !== null && delay > 0;
  }).length;

  const approvalConditionsOutstanding = section5.approvalConditions.filter(
    (condition) =>
      condition.complianceStatus === 'pending' ||
      condition.complianceStatus === 'delayed' ||
      condition.complianceStatus === 'not-sure',
  ).length;

  const snapshot =
    payload.legalUniverseMaterialityPolicyAndPartyMapping.legalDdSnapshot;

  return {
    matterCount: matters.length,
    mattersByCategory: buildMattersByCategory(matters),
    criminalMatterCount: matters.filter(isCriminalMatter).length,
    taxMatterCount: matters.filter(isTaxMatter).length,
    pendingOutcomeCount: matters.filter(
      (m) =>
        m.statusOutcome.outcomeStatus === 'pending' ||
        m.statusOutcome.outcomeStatus === 'appeal-pending' ||
        m.statusOutcome.outcomeStatus === '',
    ).length,
    exposureByCurrency: buildExposureByCurrency(matters),
    taxAggregates: buildTaxAggregates(payload),
    approvalCount: approvals.length,
    expiredApprovalCount: approvals.filter(
      (a) =>
        a.status === 'expired-renewal-applied' || a.status === 'expired-renewal-not-applied',
    ).length,
    renewalPendingCount: approvals.filter(
      (a) => a.status === 'renewal-pending' || a.status === 'application-pending',
    ).length,
    approvalExpiryWindows: buildApprovalExpiryWindows(payload, asOf),
    complianceCounts: {
      domainReviewCount: section6.complianceDomainReviews.length,
      domainsWithKnownExceptions: section6.complianceDomainReviews.filter(
        (review) => review.knownExceptions === 'yes',
      ).length,
      complianceIssueCount: section6.complianceIssues.length,
      continuingIssues: section6.complianceIssues.filter((issue) => issue.continuing === 'yes')
        .length,
      statutoryDueCount: section6.statutoryDues.length,
      delayedStatutoryDues,
      approvalConditionsOutstanding,
    },
    creditorTotals: buildCreditorTotals(payload),
    remediationOpenCount:
      payload.reconciliationRemediationAndIssuerConfirmations.remediationActions.filter(
        (action) =>
          action.status === 'open' ||
          action.status === 'in-progress' ||
          action.status === 'blocked',
      ).length,
    legalDdAsOfDate: snapshot.legalDdAsOfDate,
    reconciliation: buildReconciliationPreview(payload, linkedReferences),
  };
}
